### ISSUE TITLE
Adding / Removing Sources while a layer style still exists.

### ISSUE THREAD

**Issue by cl0ckwork** [uber/react-map-gl#474]
I've been playing around with this repo for a bit, really like the immutable aspects. I believe I've run into an issue with removing a`source` with a existing `layer(s)` on the `mapStyle` object.

In the [geojson example](https://uber.github.io/react-map-gl/#/Examples/geojson) a source and layer are added, on year change the source is updated. However, if one were to remove this example source, by taking it off the `sources` field in the `mapStyle`...

for example:
```javascript

_removeSource = () => {
    const newMapStyle = this.state.mapStyle.deleteIn(['sources', 'incomeByState']);
    this.setState({ mapStyle: newMapStyle })
  }
```
the `mapStyle` updates, but mapbox-gl does not visibly remove the layer, instead the layer remains visible. I believe this is because the `layers` are dependent on the `sources` (according to this [stackoverflow](https://stackoverflow.com/questions/42478356/removing-a-source-in-mapbox-gl-js-doesnt-remove-its-layers) answer), also while the dependent layer still exists mapbox-gl it appears as though prevents the source from being removed. Essentially `map.removeLayer` needs to be called before `map.removeSource`.

In react-map-gl, if a source is marked as `exit` in `style-utils/diffStyles`, only `map.removeSource(exit.id)` is called. Thus, the associated layers were not removed (which is why the map still renders them, from what I can gather).

so, if you were to apply the following in `style-utils/setDiffStyle` (at line 69)

```javascript
for (const exit of sourcesDiff.exit) {
    const layers = nextStyle.get('layers').toJS();
    for (const lyr of layers) {
      if (lyr.source === exit.id) {
        map.removeLayer(lyr.id);
      }
    }
    map.removeSource(exit.id);
  }
```
The map then updates with the removed layer. However, adding the source + layer back seems a bit troublesome (I am unable to do so, in my trivial example). 
```jsx
// loads the souce/layer back onto the mapStyle
<button onClick={() => this._loadData(data)}>ADD SOURCE</button>
```
This is likely because the layers weren't removed from the`mapStyle`, they are only removed from the actual `map` object. So `setDiffStyle` doesn't detect a layer change on the next go around.

Here's where it gets confusing.. if I remove the source and the layers that are associated from the `mapStyle`, everythings great. The the sources are removed, the layers are removed, and the map updates. However, when I add the source back under the same id, this shows up:

```javascript
Error: There is already a source with this ID
```
Somewhere along the line a phantom source is hanging out even though its been successfully removed from the map.. is the source still present in the map object?

**Comment by cl0ckwork** [comment]
I think I have the solution here, a two step approach:

1. Throw an error/warning if the layers were not removed with their linked sources on `mapStyle` change, something to the effect of:

```javascript
function checkForEqualLayerSourceChanges(sourceExit, nextLayers) {
  const sourceIds = sourceExit.map(s => s.id);
  const layersNotRemoved = nextLayers.filter(lyr => sourceIds.includes(lyr.get('source')));
  if (layersNotRemoved.size) {
    throw new Error(`You must remove any layers associated with sources you are removing: ${layersNotRemoved.map(l => l.get('id')).toJS().join('')}`);
  }
  return true;
}
```
2. If both layers and sources were updated then apply the removals in order (*layer first, source second*):

```javascript
  for (const enter of sourcesDiff.enter) {...}
  for (const update of sourcesDiff.update) {...}

  for (const exit of layersDiff.exiting) {
    if (map.style.getLayer(exit.id)) {
      // remove the layer before any sources are removed
      map.removeLayer(exit.id);
    }
  }
  for (const update of layersDiff.updates) {...}

  for (const exit of sourcesDiff.exit) {
    // remove the source after the associated layers are removed
    map.removeSource(exit.id);
  }
```

I have a PR I'd be happy to share if interested. Thanks!

**Comment by Pessimistress** [comment]
@cl0ckwork Thanks for digging into this. Do I understand this issue correctly:
- react-map-gl remove sources before layers in style diffing
- mapbox-gl's change in 0.43.0: [it is now an error to attempt to remove a source that is in use](https://github.com/mapbox/mapbox-gl-js/blob/master/CHANGELOG.md#0430-december-21-2017)

Feel free to open a PR of your fix.

**Comment by cl0ckwork** [comment]
@Pessimistress you are spot on, I should have included that gl change, good find there. Though that error doesn't seem to bubble up in react-map-gl, it appears to fail silently, from what I've seen.

I've created a PR #475 , feel free to provide comments, changes. thanks!

**Comment by daumann** [comment]
I seem to be having the same issue 'There is already a source with this ID' when updateIn or setIn new features of a geojson source.

**Comment by carpiediem** [comment]
I think I'm having a related issue.  I'd like to interactively change the opacity of an image layer.  But, when I modify the mapStyle property in the grandparent element (Home), `componentWillReceiveProps` is triggered in my parent element (Map), but there's no visible change to the rendered map.  I'm not altering the sources at all.

Will this functionality require a change to the source of react-map-gl?

    import React from 'react';
    import ReactMapGL from 'react-map-gl';

    import '../node_modules/mapbox-gl/dist/mapbox-gl.css';

    class Map extends React.Component {

      state = {
        viewport: {
          width: 400,
          height: 400,
          latitude: 40,
          longitude: -75,
          zoom: 3
        },
      };

      componentWillReceiveProps(nextProps) {
        // The new value of nextProps.mapStyle.layers[n].paint['raster-opacity'] is correctly logged here
        console.log(nextProps);
      }

      render() {
        return (
          <div ref="Map" className="Map">
            <ReactMapGL
              {...this.props.viewport}
              mapStyle={this.props.mapStyle}
              onViewportChange={this.props.onViewportChange}
            />
          </div>
        );
      }
    }

**Comment by Pessimistress** [comment]
This should have been fixed in v3.2.6
